sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1180 @@
+"""
+Metadata Routing Utility
+"""
+
+# Author: Adrin Jalali <adrin.jalali@gmail.com>
+# License: BSD 3 clause
+
+import inspect
+from copy import deepcopy
+from enum import Enum
+from warnings import warn
+from collections import namedtuple
+from typing import Union, Optional
+from ._bunch import Bunch
+
+# This namedtuple is used to store a (mapping, routing) pair. Mapping is a
+# MethodMapping object, and routing is the output of `get_metadata_routing`.
+# MetadataRouter stores a collection of these namedtuples.
+RouterMappingPair = namedtuple("RouterMappingPair", ["mapping", "router"])
+
+# A namedtuple storing a single method route. A collection of these namedtuples
+# is stored in a MetadataRouter.
+MethodPair = namedtuple("MethodPair", ["callee", "caller"])
```

Comment:
While perfectly valid, `callee` and `caller` differ in the last letter only. Are there good alternatives?

Reply:
They are the best we came up with (and we did go back and forth quite a bit during the development of this PR and the previous one): https://github.com/scikit-learn/scikit-learn/pull/22083#discussion_r783433393